orchestration/dagster_orchestration/hca_orchestration/resources/sam.py

```diff
@@ -0,0 +1,57 @@
+from urllib.parse import urljoin
+from cached_property import cached_property
+from dataclasses import dataclass
+
+from dagster import configured, Field, resource, StringSource
+from dagster.core.execution.context.init import InitResourceContext
+from google.auth.transport.requests import AuthorizedSession
+
+from hca_orchestration.contrib.google import authorized_session
+from hca_orchestration.support.typing import DagsterConfigDict
+
+
+@dataclass
+class Sam:
+    base_url: str
+
+    def make_snapshot_public(self, snapshot_id: str) -> None:
+        response = self._session.put(
+            self._api_url(f'api/resources/v1/datasnapshot/{snapshot_id}/policies/reader/public'),
+            data="true",  # telling the endpoint to set the flag to true
+        )
+
+        # raise an exception for a bad response
+        response.raise_for_status()
+
+    # private methods
+
+    def _api_url(self, url_part: str) -> str:
+        return urljoin(self.base_url, url_part)
+
+    @cached_property
```

Comment:
⛏️ I think we should hold off on caching/memoization until it's really necessary for performance reasons; the overhead of debugging a caching bug outweighs the performance gain we'll win from not having to auth another session.

Reply:
I'm more bullish than this on using caching for simple properties like this, where we're treating the value as immutable, but reviewing `AuthorizedSession`, it does seem to be stateful (e.g. it's closeable), so I'll ditch this in favor of directly calling the function.